Add smart answer generator

gclssvglx opened the pull request:
Add a rails generator for a new smart answer.

- Add generator
- Add usage file for use with '--help' option
- Add generator templates
- Add generator test
- Update create a new smart answer documentation

:warning: This application is Continuously Deployed: :warning:

- Merged changes are automatically deployed to staging and production.

- Make sure you follow [the guidance for deployments](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) **before** you merge.

- Check your branch is being deployed in the [Release app](https://release.publishing.service.gov.uk/applications/smartanswers), after merging.

theseanything reviewed (commented):
Think this is super handy! The boilerplate code generated doesn't seem like your typical use case though. The outcome looks like smart answers is saving your responses in database to be used for something else. I think the typical smart answer just presents a "calculated result" back to the user on the outcome page. We could change it so the outcome page body outputs the _number_ of favourite toppings. That way we'd also show how a calculator class can be used to generate a result.

reggieb commented:
I like the idea, but think the default behaviour should be simpler. There are plenty of existing flows that can act as examples, so as much as I like the idea of there being a pizza flow, it's adding unnecessary duplication.

Personally I think the default behaviour should just create a simple `Continue: yes/no`.

Perhaps if it behaved like this:
```
rails generate smart_answer foo-answer
````
Creates:
  lib/smart_answer_flows/foo-answer.rb
  lib/smart_answer_flows/foo-answer/foo_answer.erb - simple flow of example yes/no to resulting outcome
  lib/smart_answer_flows/foo-answer/questions/example.erb - (simple yes/no form)
  lib/smart_answer_flows/foo-answer/outcomes/outcome.erb - just displays something like "You chose 'Yes'"
  lib/smart_answer/calculators/foo_answer_calculator.rb

And
```
rails generate smart_answer bar
````
Creates:
  lib/smart_answer_flows/bar.rb
  lib/smart_answer_flows/bar/bar.erb
  lib/smart_answer_flows/bar/questions/example.erb
  lib/smart_answer_flows/bar/outcomes/outcome.erb
  lib/smart_answer/calculators/bar_calculator.rb
```

gclssvglx commented:
Hi @theseanything @reggieb thanks for the comments and suggestions - super helpful 👍 .  

I've updated the generator.  Hopefully it addresses your comments.

reggieb commented on doc/smart-answer-flow-development/creating-a-new-smart-answer.md:
````diff
@@ -1,8 +1,38 @@
+generator...
+
+```bash
+$ rails generate smart_answer example_smart_answer
````
I'd suggest just:
```bash
$ rails generate smart_answer my_smart_answer
```
as `example` suggests there is something special about the name.

gclssvglx replied:
Thanks @reggieb - fair point.  Also makes it consistent with line 28.

reggieb commented on doc/smart-answer-flow-development/creating-a-new-smart-answer.md:
````diff
@@ -1,8 +1,38 @@
+```bash
+$ rails generate smart_answer --help
+$ rails generate smart_answer my_smart_answer -p
+```
````
Really picky, but it might no be obvious that `-p` is the same as `--pretend`. So if you use `--pretend` in the description (lines 23-24), then `--pretend` should be used here too.

gclssvglx replied:
Thanks @reggieb - good spot.  I did see it earlier and forgot to change it 😁

reggieb commented on lib/generators/smart_answer/smart_answer_generator.rb:
```diff
@@ -0,0 +1,43 @@
+  source_root File.expand_path("templates", __dir__)
+
+  def copy_flow
+    output_filename = "lib/smart_answer_flows/#{name.dasherize}.rb"
```
Picky again - but there is only ever one file name that is parameterised in this file. So make the variable `filename`

gclssvglx replied:
Thanks @reggieb - also makes the variable name shorter 👍

reggieb reviewed (commented):
Looks good - some very picky comments added

reggieb reviewed (approved):
Looks good.
